R/utils-get_code_dependency.R

```diff
@@ -0,0 +1,498 @@
+#' @param names A `character(n)` with object names.
+#'
+#' @return `character` vector of elements of `code` calls that were required to build the side-effects and
+#' influencing objects having and impact on the `object`
+#'
+#' @keywords internal
+get_code_dependency <- function(code, names) {
+  assert_classes(code, names)
+
+  if (is_empty(code)) return(code)
+
+  code <- assert_code(code) # Turns code into expression with srcref attribute.
+  pd <- utils::getParseData(code)
+  assert_names(names, pd)
+
+  calls_pd <- extract_calls(pd)
+
+  graph <- code_graph(calls_pd)
+  indexes <- unlist(lapply(names, function(x) graph_parser(x, graph)))
+  as.character(code[unique(indexes)])
+}
+
+#' Group the result of `utils::getParseData()` into separate calls
+#' @param pd (`data.frame`) A result of `utils::getParseData()`.
+#' @keywords internal
+#' @noRd
+extract_calls <- function(pd) {
+  get_children <- function(pd, parent) {
+    idx_children <- abs(pd$parent) == parent
+    children <- pd[idx_children, c("token", "text", "id")]
+    if (nrow(children) == 0) {
+      return(NULL)
+    }
+
+    if (parent > 0) {
+      do.call(rbind, c(list(children), lapply(children$id, get_children, pd = pd)))
+    }
+  }
+  calls <- lapply(pd[pd$parent == 0, "id"], get_children, pd = pd)
+  fix_comments(calls)
+}
+
+fix_comments <- function(calls) {
+  # If the first token is a COMMENT, then it belongs to the previous call.
+  if (length(calls) >= 2) {
+    for(i in 2:length(calls)){
+      if (grepl("@linksto", calls[[i]][1, "text"])) {
+        calls[[i-1]] <- rbind(calls[[i-1]], calls[[i]][ 1, ])
+        calls[[i  ]] <-                     calls[[i]][-1, ]
+      }
+    }
+  }
+  calls
+}
+
+# code_graph ----
+
+#' Create Object Dependencies Graph Within Parsed Code
+#'
+#' @description This function constructs a dependency graph that identifies the relationships between objects in
+#' parsed code. It helps you understand which objects are needed to recreate a specific object.
+#'
+#' @param calls_pd A `list` of `data.frame`s, which is a result of `utils::getParseData()` grouped into separate calls.
+#'
+#' @return A list (of length of input `calls_pd`) where each element represents one call. Each element consists of a
+#' character vector containing names of objects that were influenced by this call, and names of objects influencing this
+#' call. Influencers appear after `":"` string, e.g. `c("a", ":", "b", "c")` means a call influenced object named `a`,
+#' and the object was influenced by objects named `b` and `c`.
+#'
+#' @keywords internal
+#' @noRd
+code_graph <- function(calls_pd) {
+
+  cooccurence <- extract_occurence(calls_pd)
+
+  side_effects <- extract_side_effects(calls_pd)
+
+  graph <- mapply(c, side_effects, cooccurence, SIMPLIFY = FALSE)
+
+  remove_graph_duplicates(graph)
+
+}
+
+#' @keywords internal
+#' @noRd
+extract_occurence <- function(calls_pd) {
```

Comment:
Please write docs of this function - what technically occurrence here means
```suggestion
#' @keywords internal
#' @noRd
extract_occurrence <- function(calls_pd) {
```